pkg/test/framework2/components/pilot/client.go

```diff
@@ -0,0 +1,122 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package pilot
+
+import (
+	"context"
+	"errors"
+	"net"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	adsapi "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"google.golang.org/grpc"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+type client struct {
+	discoveryAddr *net.TCPAddr
+	conn          *grpc.ClientConn
+	stream        adsapi.AggregatedDiscoveryService_StreamAggregatedResourcesClient
+	lastRequest   *xdsapi.DiscoveryRequest
+}
+
+func newClient(discoveryAddr *net.TCPAddr) (*client, error) {
+	conn, err := grpc.Dial(discoveryAddr.String(), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	adsClient := adsapi.NewAggregatedDiscoveryServiceClient(conn)
+	stream, err := adsClient.StreamAggregatedResources(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		conn:          conn,
+		stream:        stream,
+		discoveryAddr: discoveryAddr,
+	}, nil
+}
+
+func (c *client) CallDiscovery(req *xdsapi.DiscoveryRequest) (*xdsapi.DiscoveryResponse, error) {
+	c.lastRequest = req
+	err := c.stream.Send(req)
+	if err != nil {
+		return nil, err
+	}
+	return c.stream.Recv()
+}
+
+func (c *client) StartDiscovery(req *xdsapi.DiscoveryRequest) error {
+	c.lastRequest = req
+	err := c.stream.Send(req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *client) WatchDiscovery(timeout time.Duration,
+	accept func(*xdsapi.DiscoveryResponse) (bool, error)) error {
+	c1 := make(chan error, 1)
+	go func() {
+		for {
+			result, err := c.stream.Recv()
+			if err != nil {
+				c1 <- err
+				break
+			}
+			// ACK all responses so that when an update arrives we can receive it
+			err = c.stream.Send(&xdsapi.DiscoveryRequest{
+				Node:          c.lastRequest.Node,
+				ResponseNonce: result.Nonce,
+				VersionInfo:   result.VersionInfo,
+				TypeUrl:       c.lastRequest.TypeUrl,
+			})
+			if err != nil {
+				c1 <- err
+				break
+			}
+			accepted, err := accept(result)
```

Comment:
If Close is not called in a timely manner, then the accept lambda will get called outside the context of the test. While, accept is modeled as to return error, there is nothing stopping the author to use "t" or other context methods within the lambda. In such a case panic or undefined behavior can occur.

Reply:
We'd like to avoid excessive up-down plumbing surfacing in the test itself to handle cases like this. Should ctx.Done() be able to notify listeners so that they can trigger encapsulated shutdown handling?

Reply:
Yes. ctx.Done() is modeled, so that once it is called, all the resources allocated within that context's scope will be torn down. However, the components needs to make sure to not return control back to context before the resource state (especially the active goroutines) is torn down.

So I think what is missing here is synchronization between the goroutine here, and Close call below returning. The Close call should not return until the goroutine exits.